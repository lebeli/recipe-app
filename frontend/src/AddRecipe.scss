@import "./styles/styles.scss";

.Addrecipe {
  margin: 100px 10% 100px 10%;

  #add_recipe_container {
    #add_recipe_description {
      #add_recipe_headline {
        font-family: $primaryFont;
        font-size: 16pt;
        color: $primaryColor;
      }

      #add_recipe_button_container {
        margin-top: 20px;

        button {
          background-color: $secondaryColor;
        }

        button:hover {
          background-color: #aba931;
        }

        .MuiButton-root {
          border-radius: 0;
          box-shadow: none;
        }

        .MuiButton-label {
          font-family: $secondaryFont;
        }
      }
    }

    #add_recipe_image_container {
      text-align: center;
    }
  }
}

.AddRecipeForm {
  margin-top: 100px;
  width: 100%;

  // Close button ----------------------------

  #close_button_container {
    #close_button_description {
      font-size: 11pt;
      color: $primaryColor;
    }
    .MuiButton-root {
      color: $primaryColor;
    }
    .MuiButton-root:hover,
    .MuiButton-root:focus,
    .MuiButton-root:focus-within,
    .MuiButton-root:visited,
    .MuiButton-root:active {
      background-color: transparent;
      color: #d24e13;
    }
  }

  // Form ------------------------------------

  // Headline
  #add_recipe_form_headline {
    color: $secondaryColor;
  }

  .form_labels {
    margin-top: 30px;
    margin-bottom: 20px;
  }

  .remove_step {
    min-width: auto;

    .MuiButton-label {
      padding: 0;
    }
  }

  .MuiFormControl-root {
    width: 100%;
  }
  .recipe_time {
    width: 30%;
    min-width: fit-content;
  }

  .MuiOutlinedInput-root.Mui-focused {
    border-color: rgba(0, 0, 0, 0.23);

    .MuiOutlinedInput-notchedOutline {
      border-color: rgba(0, 0, 0, 0.87);
      border-width: 1px;
    }
  }

  .MuiFormLabel-root.Mui-focused {
    color: rgba(0, 0, 0, 0.87);
    font-family: $secondaryFont;
  }

  .MuiInputBase-input,
  .MuiInputLabel-root,
  .MuiFormLabel-root,
  .MuiTypography-root {
    font-family: $secondaryFont;
  }

  .fieldset {
    height: 20px;
  }

  // Radio buttons

  .radioButtons {
    .MuiIconButton-colorSecondary:hover {
      background-color: transparent;
    }
    .MuiRadio-colorSecondary.Mui-checked {
      color: $primaryColor;
    }
  }

  // Arbeitsschritte

  h4 {
    margin-top: 40px;
  }

  #step {
    margin-top: 0px;
  }

  .remove_step_button {
    .MuiButton-root {
      padding: 0;
    }
    .MuiButton-root:hover,
    .MuiButton-root:focus,
    .MuiButton-root:focus-within,
    .MuiButton-root:visited,
    .MuiButton-root:active {
      background-color: transparent;
      color: black;
    }
  }

  // Button for adding another working step

  .recipe_add {
    .MuiButton-root {
      color: black;
      font-family: $secondaryFont;
      text-transform: none;

      .MuiSvgIcon-root {
        margin-right: 5px;
      }
    }
    .MuiButton-root:hover,
    .MuiButton-root:focus,
    .MuiButton-root:focus-within,
    .MuiButton-root:visited,
    .MuiButton-root:active {
      background-color: transparent;
      // color: #d24e13;
      color: black;
    }
  }

  // Button for uploading an image

  #add_image_button_container {
    display: flex;
    margin-top: 20px;

    #photo_icon {
      margin-right: 10px;
    }

    .MuiButton-root {
      min-width: fit-content;
      height: fit-content;
      border-radius: 0;
      color: $secondaryColor;
      border: 1px solid $secondaryColor;
    }

    .MuiButton-root:hover {
      background-color: #aba93122;
    }

    .MuiButton-label {
      font-family: $secondaryFont;
      text-transform: none;
    }

    #filename_item {
      align-self: center;
    }
  }

  .ingredients_container,
  .steps_container {
    margin-bottom: 10px;
  }

  // Submit Button

  #recipe_form_submit_button {
    margin-top: 40px;

    .MuiButton-root {
      min-width: fit-content;
      height: fit-content;
      border-radius: 0;
      background-color: $primaryColor;
      box-shadow: none;
    }

    .MuiButton-root:hover {
      background-color: $primaryColorDarker;
    }

    .MuiButton-label {
      font-family: $secondaryFont;
      color: white;
    }
  }
}
