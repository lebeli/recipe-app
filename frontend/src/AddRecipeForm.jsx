import React, { Component } from "react";
import "./AddRecipeForm.scss";
import {
  Box,
  Grid,
  Button,
  FormControlLabel,
  Radio,
  FormLabel
} from "@material-ui/core";
import CloseIcon from "@material-ui/icons/Close";
import AddIcon from "@material-ui/icons/Add";
import TextField from "@material-ui/core/TextField";
import RadioGroup from "@material-ui/core/RadioGroup";
import PhotoIcon from "@material-ui/icons/Photo";

class AddRecipeForm extends Component {
  constructor(params) {
    super(params);
    this.state = {
      closeForm: params.closeForm,
      name: "",
      meal: "fruehstueck",
      dietForm: "keine",
      duration: "01:00",
      ingredientsAmount: 0,
      ingredients: [],
      stepsAmount: 0,
      steps: [],
      image: {}
    };

    this.addIngredient = this.addIngredient.bind(this);
    this.addStep = this.addStep.bind(this);
  }

  updateStateParameters(name, event) {
    this.setState({
      [name]: event.target.value
    });
  }

  showIngredients() {
    var ingredientsComponent = new Array();
    if (this.state.ingredientsAmount > 0) {
      for (var x = 0; x < this.state.ingredientsAmount; x++) {
        ingredientsComponent.push(
          <Grid
            className="ingredient"
            container
            spacing={2}
            key={"ingredient" + (x + 1)}
            id={"ingredient" + (x + 1)}
          >
            <Grid item sm={3}>
              <TextField
                id={x.toString()}
                className="ingredient_amount"
                required
                label="Menge"
                placeholder="Bspw. 100g"
                variant="outlined"
                onChange={evt =>
                  this.updateIngredient(evt.target.id, 1, evt.target.value)
                }
              />
            </Grid>
            <Grid item sm={8}>
              <TextField
                id={x.toString()}
                className="ingredient_name"
                required
                label="Zutat"
                placeholder="bspw. Butter"
                variant="outlined"
                onChange={evt =>
                  this.updateIngredient(evt.target.id, 0, evt.target.value)
                }
              />
            </Grid>
            {x == this.state.ingredientsAmount - 1 && (
              <Grid item xs={1} sm={1} className="remove_step_button">
                <Button
                  className="remove_step"
                  disableRipple
                  onClick={() => this.removeIngredient()}
                >
                  <CloseIcon />
                </Button>
              </Grid>
            )}
          </Grid>
        );
      }
    }
    return ingredientsComponent;
  }

  updateIngredient(ingredientsIndex, ingredientIndex, value) {
    var ingredientsCopy = Object.values(this.state.ingredients);

    var ingredient = ingredientsCopy[ingredientsIndex];
    ingredient[ingredientIndex] = value;
    ingredientsCopy.splice(ingredientsIndex, 1, ingredient);

    this.setState({
      ingredients: ingredientsCopy
    });
  }

  addIngredient() {
    // Add an entry into the ingredients array, so that it can be updated with "updateIngredient"
    var ingredientsCopy = this.state.ingredients;
    var emptyIngredient = ["", ""];
    ingredientsCopy.push(emptyIngredient);

    this.setState({
      ingredientsAmount: this.state.ingredientsAmount + 1,
      ingredients: ingredientsCopy
    });
  }

  removeIngredient() {
    var ingredientsCopy = this.state.ingredients;
    ingredientsCopy.splice(ingredientsCopy.length - 1, 1);
    this.setState({
      ingredientsAmount: this.state.ingredientsAmount - 1,
      ingredients: ingredientsCopy
    });
  }

  showSteps() {
    var steps = [];
    if (this.state.stepsAmount > 0) {
      for (var x = 0; x < this.state.stepsAmount; x++) {
        steps.push(
          <Grid
            className="step"
            container
            spacing={2}
            key={"step" + (x + 1)}
            id={"step" + (x + 1)}
          >
            <Grid item xs={10} sm={11}>
              <TextField
                id={"x" + x}
                label={"Schritt " + (x + 1)}
                variant="outlined"
                multiline
                rows="3"
                onChange={evt =>
                  this.updateStep(event.target.id, evt.target.value)
                }
              />
            </Grid>
            {x == this.state.stepsAmount - 1 && (
              <Grid item xs={1} sm={1} className="remove_step_button">
                <Button
                  className="remove_step"
                  disableRipple
                  onClick={() => this.removeStep()}
                >
                  <CloseIcon />
                </Button>
              </Grid>
            )}
          </Grid>
        );
      }
    }
    return steps;
  }

  updateStep(stepIndex, value) {
    var index = stepIndex.substring(1, stepIndex.length);
    var stepsCopy = this.state.steps;
    var step = stepsCopy[index];
    step = value;
    stepsCopy.splice(index, 1, step);

    this.setState({
      steps: stepsCopy
    });
  }

  addStep() {
    // Add an entry into the steps array, so that it can be updated with "updateStep"
    var stepsCopy = this.state.steps;
    var emptyStep = "";
    stepsCopy.push(emptyStep);

    this.setState({
      stepsAmount: this.state.stepsAmount + 1,
      steps: stepsCopy
    });
  }

  removeStep() {
    var stepsCopy = this.state.steps;
    stepsCopy.splice(stepsCopy.length - 1, 1);
    this.setState({
      stepsAmount: this.state.stepsAmount - 1,
      steps: stepsCopy
    });
  }

  handleImageUpload(file) {
    this.setState({
      file: file
    });
  }

  saveRecipe() {
    console.log("Saving recipe");
    // TODO: Api call and reset of recipe state, if saving was successful
    // Also visual feedback, if saving was successful
  }

  render() {
    return (
      <div className="AddRecipeForm">
        <Box
          display="flex"
          justifyContent="flex-end"
          id="close_button_container"
        >
          <p id="close_button_description">Rezept hinzufügen abbrechen</p>
          <Button disableRipple onClick={this.state.closeForm}>
            <CloseIcon />
          </Button>
        </Box>
        <h3 id="add_recipe_form_headline">Eigenes Rezept hinzufügen</h3>
        <form action="">
          <TextField
            id="recipe_name"
            required
            label="Rezeptname"
            variant="outlined"
            value={this.state.name}
            onChange={event => this.updateStateParameters("name", event)}
          />
          <Grid container spacing={3} id="radio_buttons_container">
            <Grid item sm={6}>
              <FormLabel component="legend" className="form_labels" required>
                Mahlzeit
              </FormLabel>
              <div className="radioButtons">
                <RadioGroup
                  aria-label="meal"
                  name="meal"
                  value={this.state.meal}
                  onChange={event => this.updateStateParameters("meal", event)}
                >
                  <FormControlLabel
                    value="fruehstueck"
                    control={<Radio disableRipple />}
                    label="Frühstück"
                  />
                  <FormControlLabel
                    value="mittagessen"
                    control={<Radio disableRipple />}
                    label="Mittagessen"
                  />
                  <FormControlLabel
                    value="abendessen"
                    control={<Radio disableRipple />}
                    label="Abendessen"
                  />
                </RadioGroup>
              </div>
            </Grid>
            <Grid item sm={6}>
              <FormLabel component="legend" className="form_labels" required>
                Ernährungsform
              </FormLabel>
              <div className="radioButtons">
                <RadioGroup
                  aria-label="diet_form"
                  name="diet_form"
                  value={this.state.dietForm}
                  onChange={event =>
                    this.updateStateParameters("dietForm", event)
                  }
                >
                  <FormControlLabel
                    value="keine"
                    control={<Radio disableRipple />}
                    label="keine"
                  />
                  <FormControlLabel
                    value="vegetarisch"
                    control={<Radio disableRipple />}
                    label="vegetarisch"
                  />
                  <FormControlLabel
                    value="vegan"
                    control={<Radio disableRipple />}
                    label="vegan"
                  />
                </RadioGroup>
              </div>
            </Grid>
          </Grid>
          <FormLabel component="legend" className="form_labels" required>
            Zubereitungsdauer
          </FormLabel>
          <TextField
            id="recipe_time"
            required
            className="recipe_time"
            label="std/min"
            type="time"
            defaultValue="01:00"
            variant="outlined"
            InputLabelProps={{
              shrink: true
            }}
            inputProps={{
              step: 300 // 5 min
            }}
            onChange={event => this.updateStateParameters("duration", event)}
          />
          <FormLabel component="legend" className="form_labels" required>
            Zutaten
          </FormLabel>
          <div className="ingredients_container">{this.showIngredients()}</div>
          <div className="recipe_add" display="flex">
            <Button
              id="add_ingredient_button"
              onClick={this.addIngredient}
              disableTouchRipple
            >
              <AddIcon />
              Zutat hinzufügen
            </Button>
          </div>
          <FormLabel component="legend" className="form_labels" required>
            Arbeitsschritte
          </FormLabel>
          <div className="steps_container">{this.showSteps()}</div>
          <div className="recipe_add" display="flex">
            <Button onClick={this.addStep} disableTouchRipple>
              <AddIcon />
              Arbeitsschritt hinzufügen
            </Button>
          </div>
          <div id="add_image_button_container">
            <Grid container spacing={1}>
              <Grid item>
                <Button variant="outlined" component="label">
                  <PhotoIcon id="photo_icon" />
                  Bild hochladen
                  <input
                    required
                    type="file"
                    style={{ display: "none" }}
                    onChange={e => this.handleImageUpload(e.target.files[0])}
                  />
                </Button>
              </Grid>
              <Grid item id="filename_item">
                <div>{this.state.file != null && this.state.file.name}</div>
              </Grid>
            </Grid>
          </div>
          <div id="recipe_form_submit_button">
            <Button variant="contained" onClick={() => this.saveRecipe()}>
              Rezept speichern
            </Button>
          </div>
        </form>
      </div>
    );
  }
}

export default AddRecipeForm;
