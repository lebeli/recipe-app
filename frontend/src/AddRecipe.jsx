import React, { Component } from "react";
import "./AddRecipe.scss";
import { Grid, Button } from "@material-ui/core";
import img from "./images/add_recipe.jpg";
import AddRecipeForm from "./AddRecipeForm";

class AddRecipe extends Component {
  constructor(params) {
    super(params);

    this.state = {
      showForm: false
    };

    this.openForm = this.openForm.bind(this);
    this.closeForm = this.closeForm.bind(this);
  }

  openForm() {
    this.setState({
      showForm: true
    });
  }

  closeForm() {
    this.setState({
      showForm: false
    });
  }

  render() {
    return (
      <div className="Addrecipe">
        <Grid container spacing={5} id="add_recipe_container">
          <React.Fragment>
            <Grid item sm={12} md={7} id="add_recipe_description">
              <h2 id="add_recipe_headline">Foodies aufgepasst!</h2>
              <p>
                Minions ipsum uuuhhh tulaliloo belloo! Wiiiii poulet tikka
                masala bananaaaa para tú jeje poopayee po kass. Pepete tank
                yuuu! Jeje jiji. Bee do bee do bee do belloo! Bananaaaa gelatooo
                bananaaaa chasy para tú tulaliloo. Jeje chasy jiji tank yuuu! Po
                kass ti aamoo! Gelatooo wiiiii jiji. Wiiiii chasy aaaaaah me
                want bananaaa! Tulaliloo po kass baboiii tank yuuu! Bananaaaa la
                bodaaa tank yuuu! Chasy hana dul sae para tú po kass poulet
                tikka masala tulaliloo poopayee belloo! Daa la bodaaa tank yuuu!{" "}
              </p>
              <p>
                Tulaliloo gelatooo poopayee wiiiii baboiii butt bananaaaa
                belloo! Poulet tikka masala baboiii chasy. Potatoooo para tú
                potatoooo tulaliloo tank yuuu! Baboiii jiji. Jeje poulet tikka
                masala uuuhhh jiji para tú pepete hana dul sae chasy. Para tú
                pepete gelatooo la bodaaa tatata bala tu tulaliloo tatata bala
                tu. Jeje hana dul sae wiiiii jeje hahaha jiji. Butt tatata bala
                tu poulet tikka masala ti aamoo! Jeje hahaha baboiii. Poopayee
                ti aamoo! Gelatooo tulaliloo bananaaaa underweaaar bananaaaa
                tulaliloo. Chasy ti aamoo! Daa para tú jeje.
              </p>
              <div id="add_recipe_button_container">
                <Button
                  id="open_form_button"
                  color="secondary"
                  variant="contained"
                  onClick={this.openForm}
                >
                  Rezept hinzufügen
                </Button>
              </div>
              {this.state.showForm && (
                <AddRecipeForm closeForm={this.closeForm} />
              )}
            </Grid>
            <Grid item sm={12} md={5} id="add_recipe_image_container">
              <img width="100%" src={img} />
            </Grid>
          </React.Fragment>
        </Grid>
      </div>
    );
  }
}

export default AddRecipe;
