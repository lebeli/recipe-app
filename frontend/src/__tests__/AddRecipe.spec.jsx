import React from "react";
import AddRecipe from "../AddRecipe";
import { shallow, mount } from "enzyme";
import AddRecipeForm from "../AddRecipeForm";
import renderer from "react-test-renderer";

describe("<AddRecipe />", () => {
  test("should render elements", () => {
    const wrapper = shallow(<AddRecipe />);
    expect(wrapper.find("#add_recipe_description"));
    expect(wrapper.find("#add_recipe_image_container"));
    expect(wrapper.find("#open_form_button"));
  });

  test("button onClick should change state", () => {
    const wrapper = shallow(<AddRecipe />);
    const button = wrapper.find("#open_form_button");
    expect(wrapper.state("showForm")).toEqual(false);
    button.simulate("click");
    expect(wrapper.state("showForm")).toEqual(true);
  });

  test("button onClick should open Form", () => {
    const wrapper = mount(<AddRecipe />);
    expect(wrapper.find(AddRecipeForm).length).toEqual(0);
    wrapper
      .find("#open_form_button")
      .at(0)
      .simulate("click");
    expect(wrapper.find(AddRecipeForm).length).toEqual(1);
    expect(wrapper.find("#recipe_name"));
    expect(wrapper.find(".radioButtons").length).toEqual(2);
    expect(wrapper.find("#recipe_time"));
    expect(wrapper.find(".ingredients_container"));
    expect(wrapper.find(".steps_container"));
    expect(wrapper.find("#add_image_button_container"));
    expect(wrapper.find("#recipe_form_submit_button"));
  });

  test("closeForm should change state", () => {
    const wrapper = shallow(<AddRecipe />);
    const instance = wrapper.instance();
    expect(wrapper.state("showForm")).toEqual(false);
    instance.openForm();
    expect(wrapper.state("showForm")).toEqual(true);
    instance.closeForm();
    expect(wrapper.state("showForm")).toEqual(false);
  });

  test("renders correctly", () => {
    const tree = renderer.create(<AddRecipe />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
