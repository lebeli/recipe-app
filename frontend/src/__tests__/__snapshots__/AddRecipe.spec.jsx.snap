// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddRecipe /> renders correctly 1`] = `
<div
  className="Addrecipe"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5"
    id="add_recipe_container"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-md-7"
      id="add_recipe_description"
    >
      <h2
        id="add_recipe_headline"
      >
        Foodies aufgepasst!
      </h2>
      <p>
        Minions ipsum uuuhhh tulaliloo belloo! Wiiiii poulet tikka masala bananaaaa para tú jeje poopayee po kass. Pepete tank yuuu! Jeje jiji. Bee do bee do bee do belloo! Bananaaaa gelatooo bananaaaa chasy para tú tulaliloo. Jeje chasy jiji tank yuuu! Po kass ti aamoo! Gelatooo wiiiii jiji. Wiiiii chasy aaaaaah me want bananaaa! Tulaliloo po kass baboiii tank yuuu! Bananaaaa la bodaaa tank yuuu! Chasy hana dul sae para tú po kass poulet tikka masala tulaliloo poopayee belloo! Daa la bodaaa tank yuuu!
         
      </p>
      <p>
        Tulaliloo gelatooo poopayee wiiiii baboiii butt bananaaaa belloo! Poulet tikka masala baboiii chasy. Potatoooo para tú potatoooo tulaliloo tank yuuu! Baboiii jiji. Jeje poulet tikka masala uuuhhh jiji para tú pepete hana dul sae chasy. Para tú pepete gelatooo la bodaaa tatata bala tu tulaliloo tatata bala tu. Jeje hana dul sae wiiiii jeje hahaha jiji. Butt tatata bala tu poulet tikka masala ti aamoo! Jeje hahaha baboiii. Poopayee ti aamoo! Gelatooo tulaliloo bananaaaa underweaaar bananaaaa tulaliloo. Chasy ti aamoo! Daa para tú jeje.
      </p>
      <div
        id="add_recipe_button_container"
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
          disabled={false}
          id="open_form_button"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="MuiButton-label"
          >
            Rezept hinzufügen
          </span>
        </button>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-md-5"
      id="add_recipe_image_container"
    >
      <img
        src="test-file-stub"
        width="100%"
      />
    </div>
  </div>
</div>
`;
